File: marketplace-service/internal/service/marketplace_service.go
package service

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"time"

	"github.com/google/uuid"
	"github.com/service-marketplace/marketplace-service/internal/domain"
)

type marketplaceService struct {
	repo domain.MarketplaceRepository
}

func NewMarketplaceService(repo domain.MarketplaceRepository) domain.MarketplaceService {
	return &marketplaceService{repo: repo}
}

func (s *marketplaceService) ListJobs(ctx context.Context, category string, lat, lng, radius float64) ([]domain.Job, error) {
	return s.repo.GetJobs(ctx, category, lat, lng, radius)
}

func (s *marketplaceService) GetJob(ctx context.Context, id string) (*domain.Job, error) {
	return s.repo.GetJobByID(ctx, id)
}

func (s *marketplaceService) PostJob(ctx context.Context, customerID string, req domain.CreateJobRequest) (*domain.Job, error) {
	job := &domain.Job{
		ID:               uuid.New().String(),
		CustomerID:       customerID,
		Title:            req.Title,
		Description:      req.Description,
		Category:         req.Category,
		Status:           "PUBLISHED",
		MaxBudget:        req.MaxBudget,
		IsEmergency:      req.IsEmergency,
		Lat:              req.Lat,
		Lng:              req.Lng,
		RecurrenceType:   req.RecurrenceType,
		TotalOccurrences: req.TotalOccurrences,
		ParentJobID:      req.ParentJobID,
		ScheduledAt:      req.ScheduledAt,
		CreatedAt:        time.Now(),
	}
	if job.ParentJobID != nil && *job.ParentJobID == "" {
		job.ParentJobID = nil
	}
	if job.TotalOccurrences == 0 {
		job.TotalOccurrences = 1
	}
	if job.RecurrenceType == "" {
		job.RecurrenceType = "ONCE"
	}
	err := s.repo.CreateJob(ctx, job)
	return job, err
}

func (s *marketplaceService) PlaceBid(ctx context.Context, providerID, jobID string, req domain.CreateBidRequest) (string, error) {
	bidID := uuid.New().String()
	bid := &domain.Bid{
		ID:            bidID,
		JobID:         jobID,
		ProviderID:    providerID,
		Amount:        req.Amount,
		EstimatedTime: req.EstimatedTime,
		Message:       req.Message,
		Status:        "PENDING",
	}
	err := s.repo.CreateBid(ctx, bid)
	return bidID, err
}

func (s *marketplaceService) ListBids(ctx context.Context, jobID string) ([]domain.Bid, error) {
	return s.repo.GetBidsByJobID(ctx, jobID)
}

func (s *marketplaceService) AcceptOffer(ctx context.Context, jobID, bidID string) error {
	return s.repo.AcceptBid(ctx, jobID, bidID)
}

func (s *marketplaceService) RejectOffer(ctx context.Context, jobID, bidID string, reason string) error {
	return s.repo.RejectBid(ctx, jobID, bidID, reason)
}

func (s *marketplaceService) CounterOffer(ctx context.Context, bidID string, userID string, req domain.CounterBidRequest) error {
	return s.repo.CounterBid(ctx, bidID, userID, req.Amount, req.Reason)
}

func (s *marketplaceService) MarkComplete(ctx context.Context, jobID, userID string, req domain.CompleteJobRequest) error {
	return s.repo.CompleteJob(ctx, jobID, userID, req.Score, req.Comment)
}

func (s *marketplaceService) ListProviderBids(ctx context.Context, providerID string) ([]domain.Bid, error) {
	return s.repo.GetBidsByProviderID(ctx, providerID)
}

func (s *marketplaceService) ListProviderJobs(ctx context.Context, providerID string) ([]domain.Job, error) {
	return s.repo.GetJobsForProvider(ctx, providerID)
}

func (s *marketplaceService) GetInsights(ctx context.Context, category string) (float64, int, error) {
	return s.repo.GetCategoryInsights(ctx, category)
}

func (s *marketplaceService) UpdateJobStatus(ctx context.Context, jobID string, status string) error {
	return s.repo.UpdateJobStatus(ctx, jobID, status)
}

func (s *marketplaceService) CancelJob(ctx context.Context, jobID string, userID string) error {
	err := s.repo.CancelJob(ctx, jobID, userID)
	if err != nil {
		return err
	}

	// Trigger refund in payment service
	paymentServiceURL := os.Getenv("PAYMENT_SERVICE_URL")
	if paymentServiceURL == "" {
		paymentServiceURL = "http://payment-service:8084"
	}

	payload := map[string]string{"job_id": jobID}
	jsonPayload, _ := json.Marshal(payload)

	req, _ := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/escrow/refund", paymentServiceURL), bytes.NewBuffer(jsonPayload))
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		// Log error but don't fail job cancellation if payment service is down
		// In production, we'd use a retry mechanism or an event bus
		fmt.Printf("Warning: Failed to trigger refund for job %s: %v\n", jobID, err)
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		fmt.Printf("Warning: Payment service returned status %d for refund of job %s\n", resp.StatusCode, jobID)
	}

	// Notify communication service
	commServiceURL := os.Getenv("COMMUNICATION_SERVICE_URL")
	if commServiceURL == "" {
		commServiceURL = "http://communication-service:8083"
	}

	commPayload := map[string]string{
		"job_id":  jobID,
		"content": "JOB_CANCELLED",
	}
	jsonCommPayload, _ := json.Marshal(commPayload)

	reqComm, _ := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/chat/system", commServiceURL), bytes.NewBuffer(jsonCommPayload))
	reqComm.Header.Set("Content-Type", "application/json")

	respComm, err := client.Do(reqComm)
	if err != nil {
		fmt.Printf("Warning: Failed to notify communication service for job %s: %v\n", jobID, err)
		return nil
	}
	defer respComm.Body.Close()

	return nil
}

File: mobile/shared_core/lib/marketplace_service.dart
import 'package:dio/dio.dart';
import 'package:latlong2/latlong.dart';
import 'network_service.dart';
import 'models.dart';

class MarketplaceService {
  final Dio _dio = NetworkService().dio;

  MarketplaceService();

  Future<List<Job>> getJobs({String? category, double? lat, double? lng, double? radius}) async {
    try {
      final response = await _dio.get('/marketplace/jobs/', queryParameters: {
        if (category != null) 'category': category,
        if (lat != null) 'lat': lat,
        if (lng != null) 'lng': lng,
        if (radius != null) 'radius': radius,
      });
      if (response.statusCode == 200) {
        final List jobsData = response.data['jobs'];
        return jobsData.map((json) => Job.fromJson(json)).toList();
      }
    } catch (e) {
      print('Get jobs error: $e');
    }
    return [];
  }

  Future<Job?> getJob(String jobId) async {
    try {
      final response = await _dio.get('/marketplace/jobs/$jobId');
      if (response.statusCode == 200) {
        return Job.fromJson(response.data);
      }
    } catch (e) {
      print('Get job error: $e');
    }
    return null;
  }

  Future<List<Bid>> getBids(String jobId) async {
    try {
      final response = await _dio.get('/marketplace/jobs/$jobId/bids');
      if (response.statusCode == 200) {
        final List bidsData = response.data['bids'];
        return bidsData.map((json) => Bid.fromJson(json)).toList();
      }
    } catch (e) {
      print('Get bids error: $e');
    }
    return [];
  }

  Future<Bid?> submitBid({
    required String jobId,
    required double amount,
    required String estimatedTime,
    String? message,
  }) async {
    try {
      final response = await _dio.post(
        '/marketplace/jobs/$jobId/bids',
        data: {
          'amount': amount,
          'estimated_time': estimatedTime,
          'message': message,
        },
      );

      if (response.statusCode == 201) {
        final id = response.data['id'];
        return Bid(
          id: id,
          jobId: jobId,
          providerId: '',
          amount: amount,
          estimatedTime: estimatedTime,
          message: message ?? '',
          status: 'pending',
          providerRating: 5.0,
          providerVerified: false,
          providerName: '',
          createdAt: DateTime.now(),
        );
      }
    } catch (e) {
      print('Submit bid error: $e');
    }
    return null;
  }

  Future<bool> acceptBid(String jobId, String bidId) async {
    try {
      final response = await _dio.post('/marketplace/jobs/$jobId/accept/$bidId');
      return response.statusCode == 200;
    } catch (e) {
      print('Accept bid error: $e');
    }
    return false;
  }

  Future<bool> rejectBid(String jobId, String bidId, {String? reason}) async {
    try {
      final response = await _dio.post(
        '/marketplace/jobs/$jobId/reject/$bidId',
        data: {'reason': reason},
      );
      return response.statusCode == 200;
    } catch (e) {
      print('Reject bid error: $e');
    }
    return false;
  }

  Future<bool> counterOffer(String bidId, double amount, {String? reason}) async {
    try {
      final response = await _dio.post(
        '/marketplace/jobs/bids/$bidId/counter',
        data: {'amount': amount, 'reason': reason},
      );
      return response.statusCode == 200;
    } catch (e) {
      print('Counter bid error: $e');
    }
    return false;
  }

  Future<bool> completeJob(String jobId, int score, String comment) async {
    try {
      final response = await _dio.post(
        '/marketplace/jobs/$jobId/complete',
        data: {'score': score, 'comment': comment},
      );
      return response.statusCode == 200;
    } catch (e) {
      print('Complete job error: $e');
    }
    return false;
  }

  Future<Map<String, dynamic>?> getInsights(String category) async {
    try {
      final response = await _dio.get('/marketplace/jobs/insights', queryParameters: {'category': category});
      if (response.statusCode == 200) {
        return response.data;
      }
    } catch (e) {
      print('Get insights error: $e');
    }
    return null;
  }

  Future<List<Bid>> getProviderBids() async {
    try {
      final response = await _dio.get('/marketplace/jobs/provider/bids');
      if (response.statusCode == 200) {
        final List bidsData = response.data['bids'];
        return bidsData.map((json) => Bid.fromJson(json)).toList();
      }
    } catch (e) {
      print('Get provider bids error: $e');
    }
    return [];
  }

  Future<List<Job>> getProviderJobs() async {
    try {
      final response = await _dio.get('/marketplace/jobs/provider/jobs');
      if (response.statusCode == 200) {
        final List jobsData = response.data['jobs'];
        return jobsData.map((json) => Job.fromJson(json)).toList();
      }
    } catch (e) {
      print('Get provider jobs error: $e');
    }
    return [];
  }

  Future<Job?> createJob({
    required String title,
    required String description,
    required String category,
    double? maxBudget,
    bool isEmergency = false,
    LatLng? location,
    RecurrenceType recurrenceType = RecurrenceType.once,
    int totalOccurrences = 1,
    String? parentJobId,
    DateTime? scheduledAt,
  }) async {
    try {
      final response = await _dio.post(
        '/marketplace/jobs/',
        data: {
          'title': title,
          'description': description,
          'category': category,
          'max_budget': maxBudget,
          'is_emergency': isEmergency,
          'lat': location?.latitude ?? 0.0,
          'lng': location?.longitude ?? 0.0,
          'recurrence_type': recurrenceType.name.toUpperCase(),
          'total_occurrences': totalOccurrences,
          'parent_job_id': parentJobId,
          'scheduled_at': scheduledAt?.toIso8601String(),
        },
      );

      if (response.statusCode == 201) {
        final json = response.data;
        return Job.fromJson(json);
      } else {
        print('Create job failed with status: ${response.statusCode}, data: ${response.data}');
      }
    } catch (e) {
      if (e is DioException) {
        print('Create job DioError: ${e.response?.statusCode} - ${e.response?.data}');
      } else {
        print('Create job error: $e');
      }
    }
    return null;
  }

  Future<bool> updateJobStatus(String jobId, JobStatus status) async {
    try {
      String statusStr = 'ACCEPTED';
      if (status == JobStatus.enRoute) {
        statusStr = 'EN_ROUTE';
      } else if (status == JobStatus.inProgress) {
        statusStr = 'IN_PROGRESS';
      } else if (status == JobStatus.completed) {
        statusStr = 'COMPLETED';
      } else if (status == JobStatus.published) {
        statusStr = 'PUBLISHED';
      }
      final response = await _dio.post(
        '/marketplace/jobs/$jobId/status',
        data: {'status': statusStr},
      );
      return response.statusCode == 200;
    } catch (e) {
      print('Update job status error: $e');
    }
    return false;
  }

  Future<bool> cancelJob(String jobId) async {
    try {
      final response = await _dio.post('/marketplace/jobs/$jobId/cancel');
      return response.statusCode == 200;
    } catch (e) {
      print('Cancel job error: $e');
    }
    return false;
  }
}
